porchlight/neighborhood.py

```diff
@@ -426,6 +457,213 @@ def call_all_doors(self):
+        req_params = input_door.arguments
+        input_params = {p: self._params[p].value for p in req_params}
+
+        args = []
+        kwargs = {}
+
+        for p, value in input_params.items():
+            if p in input_door.positional_only:
+                args.append(value)
+
+            else:
+                kwargs[p] = value
+
+        return args, kwargs
+
+    def initialize(self):
+        """Runs initialization functions present in
+        :py:attr:`~porchlight.neighborhood.Neighborhood.initialization` if
+        ``has_initialized`` is ``False`` for this
+        :py:class:`~porchlight.neighborhood.Neighborhood`.
+        """
+        # Do nothing if initialization has already happened.
+        if self.has_initialized or not self.initialization:
+            return
+
+        # Ensure initialization is iterable, if not
+        if not hasattr(self.initialization, "__iter__"):
+            self.initialization = [self.initialization]
+
+            # Log this any time initialization happens, because this is
+            # important.
+            logging.info(
+                f"Could not iterate through object: "
+                f"{type(self.initialization)}, will now attempt to call "
+                f"the object."
+            )
+
+        # Ensure that none of the doors are DynamicDoors, as this can cause
+        # unexpected behaviour.
+        if any(isinstance(f, door.DynamicDoor) for f in self.initialization):
+            msg = "DynamicDoors are not yet supported for initialization."
+
+            logging.error(msg)
+            raise NotImplementedError(msg)
+
+        # Call each of the callables in the initialization function. If they
+        # are doors, save their parameter states.
+        for fxn in self.initialization:
+            if not isinstance(fxn, door.Door):
+                fxn = door.Door(fxn)
+
+            return_values = fxn.return_vals
+            arguments, keyword_arguments = self.gather_door_arguments(fxn)
+
+            result = fxn(*arguments, **keyword_arguments)
+
+            # Update any parameters already present. For initialization, only
+            # recognized return values are updated.
+            #
+            # TODO: ^ This behaviour needs to be better documented.
+            neighborhood_params = tuple(self.params.keys())
+
+            if len(return_values) == 1:
+                result = [result]
+
+            for retval, value in zip(return_values, result):
+                if retval in neighborhood_params:
+                    self.set_param(retval, value)
+
+    def finalize(self):
+        """Finalization executes doors/callables found in
+        ``Neighborhood.finalization``. It must be invoked directly by the user.
+
+        Unlike initialization, finalization will add new constant
+        :py:class:`~porchlight.param.Param`s to the ``Neighborhood`` object.
+        """
+        # Ensure finalization is iterable, if not
```

Comment:
Expand comment

Reply:
Addressed in https://github.com/teald/porchlight/pull/66/commits/3c8793be79e96e455a7e778718a6e5405cebbbbf